ConfigSpace/c_util.pyx

```diff
@@ -121,110 +121,48 @@ cpdef np.ndarray correct_sampled_array(
-    int num_hyperparameters,
-    list unconditional_hyperparameters,
+    list conditional_hyperparameters,
     dict hyperparameter_to_idx,
     dict parent_conditions_of,
-    dict parents_of,
-    dict children_of,
 ):
+    """Ensure that the array values of inactive hyperparameters are NaN.
+    
+    The output array does not violate any condition or forbidden clause.
+
+    Parameters
+    ----------
+    vector : np.ndarray
+        Vector of hyperparameter values. It is assumed that none of the active hyperparameters has a NaN value assigned.
+    
+    conditional_hyperparameters : list[str]
+        Names of conditional hyperparameters ordered topologically
+
+    Returns
+    -------
+    np.ndarray
+        Updated vector
+    """
     cdef AbstractForbiddenComponent clause
     cdef ConditionComponent condition
-    cdef int hyperparameter_idx
     cdef DTYPE_t NaN = np.NaN
-    cdef set visited
-    cdef set inactive
-    cdef Hyperparameter child
-    cdef list children
-    cdef str child_name
-    cdef list parents
-    cdef Hyperparameter parent
-    cdef int parents_visited
-    cdef list conditions
-    cdef int add
-
-    cdef int* active
-    active = <int*> malloc(sizeof(int) * num_hyperparameters)
-    for j in range(num_hyperparameters):
-        active[j] = 0
+    cdef str current_name
 
     for j in range(len(forbidden_clauses_unconditionals)):
         clause = forbidden_clauses_unconditionals[j]
         if clause.c_is_forbidden_vector(vector, strict=False):
-            free(active)
             raise ForbiddenValueError(
                 "Given vector violates forbidden clause %s" % (
                     str(clause)
                 )
             )
+    
+    # We assume that the conditional hyperparameters are ordered in topological order.
+    for current_name in conditional_hyperparameters:
+        for condition in parent_conditions_of[current_name]:
+            if not condition._evaluate_vector(vector):
+                vector[hyperparameter_to_idx[current_name]] = NaN
+                break            
 
-    hps = deque()
-    visited = set()
-    hps.extendleft(hyperparameters_with_children)
-
-    for ch in unconditional_hyperparameters:
-        active[hyperparameter_to_idx[ch]] = 1
-
-    inactive = set()
-
-    while len(hps) > 0:
-        hp = hps.pop()
-        visited.add(hp)
-        children = children_of[hp]
-        for child in children:
-            child_name = child.name
-            if child_name not in inactive:
-                parents = parents_of[child_name]
-                hyperparameter_idx = hyperparameter_to_idx[child_name]
-                if len(parents) == 1:
-                    conditions = parent_conditions_of[child_name]
-                    add = True
-                    for j in range(len(conditions)):
-                        condition = conditions[j]
-                        if not condition._evaluate_vector(vector):
-                            add = False
```

Comment:
This may be too early to declare `child` inactive. We know at this point that `child` has 1 parent and that at least one of its parent conditions is not satisfied. However, if `conditions` are in disjunction, the `child` still has a chance of being active.